module/applications/components/damage-application.mjs

```diff
@@ -0,0 +1,334 @@
+    this.chatMessage = game.messages.get(messageId);
+    if ( !this.chatMessage ) return;
+
+    // Build the frame HTML only once
+    if ( !this.targetList ) {
+      const div = document.createElement("div");
+      div.classList.add("card-tray", "damage-tray", "collapsible");
+      div.innerHTML = `
+        <label class="roboto-upper">
+          <i class="fa-solid fa-heart-crack"></i>
+          <span>${game.i18n.localize("DND5E.Apply")}</span>
+          <i class="fa-solid fa-caret-down"></i>
+        </label>
+        <div class="collapsible-content">
+          <div class="target-source-control">
+            <button type="button" class="unbutton" data-mode="targeted" aria-pressed="true">
+              <i class="fa-solid fa-bullseye" inert></i> ${game.i18n.localize("DND5E.Tokens.Targeted")}
+            </button>
+            <button type="button" class="unbutton" data-mode="selected" aria-pressed="false">
+              <i class="fa-solid fa-expand inert"></i> ${game.i18n.localize("DND5E.Tokens.Selected")}
+            </button>
+          </div>
+          <ul class="targets unlist"></ul>
+          <button class="apply-damage" type="button" data-action="applyDamage">
+            <i class="fa-solid fa-reply-all fa-flip-horizontal inert"></i>
+            ${game.i18n.localize("DND5E.Apply")}
+          </button>
+        </div>
+      `;
+      this.replaceChildren(div);
+      this.applyButton = div.querySelector(".apply-damage");
+      this.applyButton.addEventListener("click", this._onApplyDamage.bind(this));
+      this.targetList = div.querySelector(".targets");
+      this.targetSourceControl = this.querySelector(".target-source-control");
+      this.targetSourceControl.querySelectorAll("button").forEach(b =>
+        b.addEventListener("click", this._onChangeTargetMode.bind(this))
+      );
+      if ( !game.user.isGM || !this.chatMessage.getFlag("dnd5e", "targets")?.length ) {
+        this.targetSourceControl.hidden = true;
+      }
+      div.querySelector(".collapsible-content").addEventListener("click", event => {
+        event.stopImmediatePropagation();
+      });
+    }
+
+    this.buildTargetsList();
+  }
+
+  /* -------------------------------------------- */
+
+  /**
+   * Build a list of targeted tokens based on current mode & replace any existing targets.
+   */
+  async buildTargetsList() {
+    // Ensure collapsible section automatically resizes when content changes
+    const collapsible = this.querySelector(".collapsible-content");
+    if ( collapsible.style.height !== "0px" ) collapsible.style.height = "auto";
+
+    let targetedTokens;
+    switch ( this.targetingMode ) {
+      case "targeted":
+        targetedTokens = (this.chatMessage.getFlag("dnd5e", "targets") ?? []).map(t => t.uuid);
+        if ( this.selectedTokensHook ) {
+          Hooks.off("controlToken", this.selectedTokensHook);
+          this.selectedTokensHook = null;
+        }
```

Comment:
Managing the hooks feels more appropriate to perform in the `targetingMode` setter. Perhaps we could move it there and then have `connectedCallback` invoke `this.targetingMode = this.targetingMode` instead of `this.buildTargetsList()` (and we would also need to make both controls `aria-pressed="false"` initially)? Now that I've written it, it looks a bit silly, but perhaps worth considering at least.

Reply:
Our linter complains about assigning a variable to itself, but I can easily determine what the initial mode should be based on the current state.